### PR TITLE
[ISSUE #8429] Fix trace message loss when traffic is heavy

### LINKED ISSUE #8429: [Bug] When the client tps is high, trace messages will be lost
### Before Creating the Bug Report

- [X] I found a bug, not just asking a question, which should be created in [GitHub Discussions](https://github.com/apache/rocketmq/discussions).

- [X] I have searched the [GitHub Issues](https://github.com/apache/rocketmq/issues) and [GitHub Discussions](https://github.com/apache/rocketmq/discussions)  of this repository and believe that this is not a duplicate.

- [X] I have confirmed that this bug belongs to the current repository, not other repositories of RocketMQ.

### Runtime platform environment

all

### RocketMQ version

all

### JDK Version

all

### Describe the Bug

当客户端消息生成或者消费tps较高时，轨迹消息batch量会比较大，由于之前没限制batch size，当batch size太大时候会出现轨迹消息概率性丢失
When the TPS (Transactions Per Second) of client message generation or consumption is high, the batch size of trajectory messages can become quite large. Due to the lack of a batch size limit previously, when the batch size becomes too large, there is a probabilistic loss of trajectory messages.

### Steps to Reproduce

1. 限制batch siez
2. 部分重构客户端发送轨迹消息的逻辑

1. Limit batch size
2. Partially refactor the client's logic for sending trajectory messages

### What Did You Expect to See?

nothing

### What Did You See Instead?

nothing

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by LetLetMe**
### Which Issue(s) This PR Fixes

Fixes #8429

### Brief Description

当客户端消息生成或者消费tps较高时，轨迹消息batch量会比较大，由于之前没限制batch size，当batch size太大时候会出现轨迹消息丢失

When the TPS (Transactions Per Second) of client message generation or consumption is high, the batch size of trajectory messages can become quite large. Due to the lack of a batch size limit previously, when the batch size becomes too large, there is a probabilistic loss of trajectory messages.

### CODE REVIEW

File: client/src/main/java/org/apache/rocketmq/client/trace/AsyncTraceDispatcher.java [hunk 1]
```diff
@@ -44,55 +44,55 @@
...
     private final long pollingTimeMil;
     private final long waitTimeThresholdMil;
     private final DefaultMQProducer traceProducer;
-    private final ThreadPoolExecutor traceExecutor;
-    // The last discard number of log
     private AtomicLong discardCount;
     private Thread worker;
+    private final ThreadPoolExecutor traceExecutor;
+
+    private final int threadNum = Math.max(8, Runtime.getRuntime().availableProcessors());
```
Comment (lizhimins) [inline client/src/main/java/org/apache/rocketmq/client/trace/AsyncTraceDispatcher.java:69]: 要这么多线程吗

File: client/src/main/java/org/apache/rocketmq/client/trace/AsyncTraceDispatcher.java [hunk 2]
```diff
@@ -197,26 +198,14 @@ public boolean append(final Object ctx) {
...
             try {
-                Thread.sleep(1);
-            } catch (InterruptedException e) {
-                break;
+                flushTraceContext();
+            } catch (Throwable e) {
             }
         }
-        log.info("------end trace send " + traceContextQueue.size() + "   " + appenderQueue.size());
+        log.warn("There are still some traces that haven't been sent " + traceContextQueue.size() + "   " + appenderQueue.size());
```
Comment (lollipopjin) [inline client/src/main/java/org/apache/rocketmq/client/trace/AsyncTraceDispatcher.java:208]: no need to print warn log here if  traceContextQueue's size is 0.

File: client/src/main/java/org/apache/rocketmq/client/trace/AsyncTraceDispatcher.java [hunk 3]
```diff
@@ -44,55 +44,55 @@
...
     private String group;
     private Type type;
     private String namespaceV2;
 
-    public AsyncTraceDispatcher(String group, Type type, String traceTopicName, RPCHook rpcHook) {
+    public AsyncTraceDispatcher(String group, Type type, int batchSize, String traceTopicName, RPCHook rpcHook) {
         // queueSize is greater than or equal to the n power of 2 of value
         this.queueSize = 2048;
-        this.batchSize = 100;
+        this.batchSize = Math.min(batchSize, 20);
```
Comment (drpmma) [inline client/src/main/java/org/apache/rocketmq/client/trace/AsyncTraceDispatcher.java:88]: It's recommended to add some annotations about `min`.

File: client/src/main/java/org/apache/rocketmq/client/trace/AsyncTraceDispatcher.java [hunk 4]
```diff
@@ -258,146 +247,96 @@ public void removeShutdownHook() {
...
 
-        private void sendDataByTimeThreshold() {
-            long now = System.currentTimeMillis();
-            for (TraceDataSegment taskInfo : taskQueueByTopic.values()) {
-                if (now - taskInfo.firstBeanAddTime >= waitTimeThresholdMil) {
-                    taskInfo.sendAllData();
+    private void flushTraceContext() {
+        List<TraceContext> contextList = new ArrayList<>(batchSize);
+        int size = traceContextQueue.size();
+        if (size != 0 && (size >= batchSize || System.currentTimeMillis() - lastFlushTime > 5000)) {
```
Comment (drpmma) [inline client/src/main/java/org/apache/rocketmq/client/trace/AsyncTraceDispatcher.java:268]: It's better to use a var rather than magic number 5000.

**Review by lizhimins — DISMISSED** [review]

File: client/src/main/java/org/apache/rocketmq/client/trace/AsyncTraceDispatcher.java [hunk 5]
```diff
@@ -197,37 +191,28 @@ public boolean append(final Object ctx) {
...
-                break;
+                flushTraceContext(true);
+            } catch (Throwable throwable) {
+                log.error("flushTraceContext error", throwable);
             }
         }
-        log.info("------end trace send " + traceContextQueue.size() + "   " + appenderQueue.size());
+        if (appenderQueue.size() > 0) {
+            log.error("There are still some traces that haven't been sent " + traceContextQueue.size() + "   " + appenderQueue.size());
+        }
```
Comment (lizhimins) [inline client/src/main/java/org/apache/rocketmq/client/trace/AsyncTraceDispatcher.java:204]: 这里有并发问题，刚 flush 完就有数据进来了

**Review by lizhimins — APPROVED** [review]

**Pull request merged**